compiler/modules/print.wasm/relooper.bal

```diff
@@ -1,37 +1,482 @@
+    private int shapeId = 1;
 
     function init(Module module) {
-
+        self.module = module;
     }
 
     function addBranch(Block fromB, Block to, Expression? condition = ()) {
-        panic error("unimplemented");
+        Branch branch = {
+            condition: condition
+        };
+        BlockBranchMap blockBranch = {
+            block: to,
+            branch: branch
+        };
+        fromB.branchesOut.push(blockBranch);
+        to.branchesIn.push(fromB);
     }
 
     function addBlock(Expression code) returns Block {
-        panic error("unimplemented");        
+        Block block = {
+            code: code,
+            id: self.blockId
+        };
+        self.blockId = self.blockId + 1;
+        self.blocks.push(block);
+        return block;
+    }
+
+    function checkExistence(Block[]? blocks, int id) returns boolean {
+        if blocks != () {
+            foreach Block b in blocks {
+                if b.id == id {
+                    return true;
+                }
+            }
+            return false;
+        }
+        return false;
+    } 
+
+    function createSimpleShape(Block[] blocks, Block entry) returns Shape {
+        Block[] nextEntries = [];
+        foreach BlockBranchMap next in entry.branchesOut {
+            if self.checkExistence(blocks, next.block.id) {
+                nextEntries.push(next.block);
+            }
+            next.block.branchesIn = next.block.branchesIn.filter(b => b.id != entry.id);
+        }
+        Block[] validBlocks = blocks.filter(b => b.id != entry.id);
+        SimpleShape simple = {
+            inner: entry,
+            next: self.calculate(validBlocks,nextEntries)
+        };
+        return simple;
+    }
+
+    function createLoopShape(Block[] blocks, Block[] entries) returns Shape {
+        Block[] nextEntries = [];
+        Block[] innerBlocks = [];
+        Block[] validBlocks = blocks.clone();
+        foreach Block entry in entries {
+            foreach Block prev in entry.branchesIn {
+                if innerBlocks.indexOf(prev) == () {
+                    innerBlocks.push(prev);
+                    validBlocks = blocks.filter(b => b.id != prev.id);
+                }
+            }    
+        }
+        foreach Block inner in innerBlocks {
+            foreach BlockBranchMap block in inner.branchesOut {
+                if innerBlocks.indexOf(block.block) == () {
+                    nextEntries.push(block.block);
+                }
+            }
+        }
+        LoopShape loop = {
+            inner: self.calculate(innerBlocks, entries),
+            next : self.calculate(validBlocks, nextEntries)
+        };
+        return loop;
+    }
+
+    function invalidateChildren(Block block, map<Block?> ownership, map<Block[]> independentGroups) {
+        Block[] queue = [block];
+        while queue.length() > 0 {
+            Block curr = queue.remove(0);
+            int? index = ownership.keys().indexOf(curr.id.toString());
```

Comment:
Can we do `ownership[curr.id.toString()]` and check if that's nil? (it might be even better to use table)

Reply:
As I'm aware key of the table should an attribute inside the record. In this case owner's `id` is not an attribute in the record. That was why a table was not used.